### ISSUE TITLE
Can't build Pillow on Debian armhf

### ISSUE THREAD

**Issue by bochecha** [python-pillow/Pillow#2851]
### What did you do?

On a Debian armhf machine, I tried running: (keeping only the relevant parts)

```
$ python setup.py install
...
Looking for zlib
Checking for include file zlib.h in /tmp/pip-rje8ft8s-build/libImaging
Checking for include file zlib.h in /usr/include
Found zlib.h
Couldn't find library z in ['/usr/lib', '/usr/lib']
...
The headers or library files could not be found for zlib,
a required dependency when compiling Pillow from source.

Please see the install instructions at:
   https://pillow.readthedocs.io/en/latest/installation.html

Traceback (most recent call last):
...
```

The problem is that [the code which detects the library dirs to search in](https://github.com/python-pillow/Pillow/blob/master/setup.py#L313-L341) just doesn't work at all on Debian.

It suffers from many problems, but let's start with the ARM-related protion:

```python
                    elif platform_ in ["arm", "armv7l"]:
                        _add_directory(library_dirs,
                                       "/usr/lib/arm-linux-gnueabi")
                        break
```

This is missing the `/usr/lib/arm-linux-gnueabihf` directory for the armhf architecture (ARM 32 bits with hardware floating point)

More fundamentally, `platform_` in this case comes from `arch_tp = (plat.processor(), plat.architecture()[0])`. On a Debian ARM system, that's `('', '32bit')`, which means we never even pass in the above condition anyway.

But in fact, this whole architecture detection code suffers from the same problem, because even on a less exotic system, for example x86_64, then `arch_tp = ('', '64bit')`. Basically Python's `platform.processor()` always seems to return an empty string on Debian.

This means that the conditional below can never evaluate to `True`, even in the case it was designed for:

```python
            if arch_tp == ("x86_64", "32bit"):
                # 32-bit build on 64-bit machine.
                _add_directory(library_dirs, "/usr/lib/i386-linux-gnu")
```

The only reason Pillow builds at all on Debian x86_64 and i386 is because those two conditions are two broad:

```python
                    if platform_ in ["x86_64", "64bit"]:
                        _add_directory(library_dirs, "/lib64")
                        _add_directory(library_dirs, "/usr/lib64")
                        _add_directory(library_dirs,
                                       "/usr/lib/x86_64-linux-gnu")
                        break
                    elif platform_ in ["i386", "i686", "32bit"]:
                        _add_directory(library_dirs, "/usr/lib/i386-linux-gnu")
                        break
```

The `64bit` and `32bit` also match AArch64 and ARM machines, where the above multiarch libdirs really shouldn't be added as they are the wrong ones.

---

For reference, the Debian package of Pillow uses this patch (among others) : 

```patch
diff --git a/setup.py b/setup.py
index 4948faf..02676ba 100755
--- a/setup.py
+++ b/setup.py
@@ -311,61 +311,6 @@ class pil_build_ext(build_ext):
                 _add_directory(include_dirs, "/usr/X11/include")

         elif sys.platform.startswith("linux"):
-            arch_tp = (plat.processor(), plat.architecture()[0])
-            if arch_tp == ("x86_64", "32bit"):
-                # 32-bit build on 64-bit machine.
-                _add_directory(library_dirs, "/usr/lib/i386-linux-gnu")
-            else:
-                for platform_ in arch_tp:
-
-                    if not platform_:
-                        continue
-
-                    if platform_ in ["x86_64", "64bit"]:
-                        _add_directory(library_dirs, "/lib64")
-                        _add_directory(library_dirs, "/usr/lib64")
-                        _add_directory(library_dirs,
-                                       "/usr/lib/x86_64-linux-gnu")
-                        break
-                    elif platform_ in ["i386", "i686", "32bit"]:
-                        _add_directory(library_dirs, "/usr/lib/i386-linux-gnu")
-                        break
-                    elif platform_ in ["aarch64"]:
-                        _add_directory(library_dirs, "/usr/lib64")
-                        _add_directory(library_dirs,
-                                       "/usr/lib/aarch64-linux-gnu")
-                        break
-                    elif platform_ in ["arm", "armv7l"]:
-                        _add_directory(library_dirs,
-                                       "/usr/lib/arm-linux-gnueabi")
-                        break
-                    elif platform_ in ["ppc64"]:
-                        _add_directory(library_dirs, "/usr/lib64")
-                        _add_directory(library_dirs,
-                                       "/usr/lib/ppc64-linux-gnu")
-                        _add_directory(library_dirs,
-                                       "/usr/lib/powerpc64-linux-gnu")
-                        break
-                    elif platform_ in ["ppc"]:
-                        _add_directory(library_dirs, "/usr/lib/ppc-linux-gnu")
-                        _add_directory(library_dirs,
-                                       "/usr/lib/powerpc-linux-gnu")
-                        break
-                    elif platform_ in ["s390x"]:
-                        _add_directory(library_dirs, "/usr/lib64")
-                        _add_directory(library_dirs,
-                                       "/usr/lib/s390x-linux-gnu")
-                        break
-                    elif platform_ in ["s390"]:
-                        _add_directory(library_dirs, "/usr/lib/s390-linux-gnu")
-                        break
-                else:
-                    raise ValueError(
-                        "Unable to identify Linux platform: `%s`" % platform_)
-
-                # XXX Kludge. Above /\ we brute force support multiarch. Here we
-                # try Barry's more general approach. Afterward, something should
-                # work ;-)
                 self.add_multiarch_paths()

                 # termux support for android.
```

That's right, they just completely remove the attempt at automatically detecting things, and directly use the `dpkg-architecture` code path (in `self.add_multiarch_paths()`), which otherwise gets skipped by the bad conditionals.

It seems like this part of `setup.py` needs to be reworked. 😕

**Comment by wiredfool** [comment]
You know, it would be helpful if Debian wouldn't silently patch stuff like this and not file some sort of upstream bug.

**Comment by wiredfool** [comment]
Most of that platform detection code is from redhat, and should match their library directory layouts. 

This is all a hairy ball of mess because the ground truth of what installs where is so hard to continually test. We could unit test parts of it, but we'd wind up having to mock things like debian not returning anything in `platform.processor()`, which ends up not telling us much.  We don't have access to the more exotic machines, and we don't have CI running on anything but x86_64 and a couple of x86 docker images.  So we end up trying to tightly tailor cases so as to not break other working installs.

I've pushed a promotion of the multiarch paths test, and then factored the long set of elifs into a data structure. 

There's got to be a better way, but that better way is likely different on each distro.

**Comment by bochecha** [comment]
> There's got to be a better way,

On Linux (and UNIX I guess?), that better way is `pkgconfig`. All the dependencies of Pillow (I checked earlier today) have the required `*.pc` files.

However it's completely reversed from what is currently done in the setup.py:

* the current code tries to add all the directories, then search for the libs/headers in those directories; all those directories are added as `include_dirs` (`-I` flag) and `library_dirs` (`-L` flag), whether they actually contain a lib that was found or not;
* the way things usually work is that we use pkgconfig to search for the libs, which returns the directories to pass as `include_dirs` and `library_dirs` for the found libs;

Basically, on Linux what should be done is something like (extremely simplified):

```python
def pkgconfig(*libs):
    flags = {'-I': 'include_dirs', '-L': 'library_dirs', '-l': 'libraries'}
    result = {'include_dirs': [], 'library_dirs': [], 'libraries': []}
    output = subprocess.check_output(['pkg-config', '--libs', '--cflags', ' '.join(libs).strip()

    for token in output.split():
        result[flags[token[:2]]].append(token[2:])

    return result

setup(
    ...
    ext_modules=[
        Extension(
            'extension',
            [
                'extension_main.c',
            ],
            **pkgconfig('zlib', 'libjpeg', 'freetype2')
        ),
    ],
)
```

**Comment by wiredfool** [comment]
Most packages have pkgconfig definitions, but there are certainly distros that don't have or ship pkgconfig by default. (e.g., ubuntu 14.04 and centos 6 to name two that I have shells open on right now)

**Comment by gerardgilliland** [comment]
Thank you for submitting this issue.
I have the same problem on a Raspberry Pi 3-B and left a comment at:
 https://github.com/python-pillow/Pillow/pull/2853
referencing this page.

**Comment by bochecha** [comment]
> there are certainly distros that don't have or ship pkgconfig by default. (e.g., ubuntu 14.04 and centos 6 to name two that I have shells open on right now)

But then they also don't ship the development headers for those libraries by default, and neither do they ship a compiler by default. Those must be installed before installing Pillow.

pkgconfig is **the** standard way to find libraries on Linux (and UNIX?). Everything relies on it exclusively since forever.

**Comment by hugovk** [comment]
@bochecha Please could you submit a PR using pkgconfig so we can compare and test the two approaches?

**Comment by bochecha** [comment]
> Please could you submit a PR using pkgconfig so we can compare and test the two approaches?

I can try. I won't guarantee when it's ready though, as I'm travelling at the moment.

However, I have access to nothing else but Linux, so I'm a bit worried of breaking other platforms, and I won't be able to test on them.

**Comment by wiredfool** [comment]
Ok, I was wrong that they don't ship pkg-config, it's that they don't ship complete pkg-config. 

@bochecha This is on centos-6, on the current docker file:

```
[pillow@53aa95e47760 /]$ yum provides */pkg-config
...
1:pkgconfig-0.23-9.1.el6.x86_64 : A tool for determining compilation options
Repo        : installed
Matched from:
Filename    : /usr/bin/pkg-config

[pillow@53aa95e47760 /]$ pkg-config --libs --cflags libjpeg
Package libjpeg was not found in the pkg-config search path.
Perhaps you should add the directory containing `libjpeg.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libjpeg' found

[pillow@53aa95e47760 /]$ pkg-config --list-all | grep jpeg
[pillow@53aa95e47760 /]$ pkg-config --list-all | grep tiff
[pillow@53aa95e47760 /]$ pkg-config --list-all | grep freetype
freetype2        FreeType 2 - A free, high-quality, and portable font engine.
[pillow@53aa95e47760 /]$ pkg-config --list-all | grep zlib
zlib             zlib - zlib compression library
```

Pkg-config is installed on centos, but their packages don't include pkg-config files for all of the libraries. 
Amazon Linux doesn't include pkg-config for libjpeg, but they do get tiff. Same with Ubuntu 14.04.

**Comment by bochecha** [comment]
Ah indeed, the EL6 package might be older than the time when pkgconfig support was added in libjpeg-turbo upstream. 😞

**Comment by hugovk** [comment]
Well, let's go with PR https://github.com/python-pillow/Pillow/pull/2853 for now as a step in the right direction, and future improvements are always welcome.

**Comment by bochecha** [comment]
I confirm #2853 fixes the build for me on Debian/armhf. Thanks a lot for that!

I still believe using pkgconfig would be better in the long run, but as I mentioned I'm not sure when I'll have the time to work on that. Since the problem is fixed though, it's not really urgent any more though.

When is the next release due? 🙂

**Comment by hugovk** [comment]
Releases are quarterly, next Pillow 4.4.0 is due 2018-01-01.
